### PR TITLE
New Mob [Garden Gnomes]

### PR DESCRIPTION
**Pull request by Comxy**
## About The Pull Request
Garden gnomes can go into the ground when they stand still for a while. If they are damaged and underground they will heal. If they are underground they will also gain damage resistance. They can make people trip when underground and deal some damage. They are available through golden slime cores or you can collect them from cargo with an emag. If one of them gets damaged they will all retaliate against the attacker.

They have a very realistic colour pallet based on online gathered data. Depending on their colour pallets some of them have a very low chance of spawning.
## Why It's Good For The Game
These new garden gnomes come with expanding the world building of space station 13. These gnomes will add more variety to the game play and use an interesting AI and behavior that interacts with the player.
gnome_promo1
gnome_promo2
## Changelog
:cl:
add: Adds garden gnomes.
imageadd: Adds garden gnome sprites.
/:cl:

### CODE REVIEW

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by Iamgoofball** [comment]
@MrMelbert is that not sufficient?

**Comment by jlsnow301** [comment]
>not a botany thing
>can't place garden gnomes in the garden I grew
>2563 -> 1984

**Comment by necromanceranne** [comment]
22fdb7169040e1da21302282eb0b9cae078a58b8

**Comment by Comxy** [comment]
I'll to look at the coding stuff later, also inhand sprites would need to be gags which sounds like a hassle >_>

**Comment by dragomagol** [comment]
Love it

**Comment by optimumtact** [comment]
maintainers add this

**Comment by Flurrieon** [comment]
i'gm fging gnomig

**Comment by Wallemations** [comment]
Delete this PR before the government eliminates us for propagating gnomes!!!

**Comment by ZeroKelvn** [comment]
Oh they're not marked as Contraband, they're... actually hidden? The only crate currently locked behind an emag, as opposed to hacking a console to unlock Contraband, is the Special Ops Crate, which contains incredibly destructive equipment, including a Traitor item. Are these gnomes really at such a similar level of power to warrant an Emag instead of Contraband?

**Comment by Sealed101** [comment]
> Oh they're not marked as Contraband, they're... actually hidden?
they're gnomes, what do you think

**Comment by JohnFulpWillard** [comment]
> Oh they're not marked as Contraband, they're... actually hidden? The only crate currently locked behind an emag, as opposed to hacking a console to unlock Contraband, is the Special Ops Crate, which contains incredibly destructive equipment (EMP and Incendiary nades, 3 smoke bombs, and a sleepypen). Are these gnomes really at such a similar level of power to warrant an Emag instead of Contraband?

https://youtu.be/4HI8prOyvw4?t=30

**Comment by MMMiracles** [comment]
> Oh they're not marked as Contraband, they're... actually hidden? The only crate currently locked behind an emag, as opposed to hacking a console to unlock Contraband, is the Special Ops Crate, which contains incredibly destructive equipment (EMP and Incendiary nades, 3 smoke bombs, and a sleepypen). Are these gnomes really at such a similar level of power to warrant an Emag instead of Contraband?

gnome hands typed this post

File: code/datums/elements/ai_retaliate.dm [hunk 1]
```diff
@@ -3,12 +3,18 @@
  * The AI controller is responsible for doing anything with that information.
  */
 /datum/element/ai_retaliate
+	/// What mob type the parent calls
+	var/mob/living/mob_type
+	/// In what range the parent calls the same mobs to retaliate
```
Comment (CapybaraExtravagante) [inline code/datums/elements/ai_retaliate.dm:8]: This should not be handled by the retaliate component but by the AI itself. (E.g. "If I have enemies in my retaliate list, I should call for help")

File: code/datums/elements/ai_retaliate.dm [hunk 2]
```diff
@@ -27,3 +33,7 @@
 		enemy_refs = list()
 	enemy_refs |= WEAKREF(attacker)
 	victim.ai_controller.blackboard[BB_BASIC_MOB_RETALIATE_LIST] = enemy_refs
+	if(mob_type)
+		for (var/mob/living/potential_victim in oview(victim, call_range))
+			if (istype(potential_victim, mob_type) || !potential_victim.stat == DEAD)
+				potential_victim.ai_controller.blackboard[BB_BASIC_MOB_RETALIATE_LIST] = enemy_refs
```
[the remark this answers is not in the document]
Comment (CapybaraExtravagante) [reply code/datums/elements/ai_retaliate.dm:39]: It's making nearby mobs of the same type also aggressive. so if you hit one gnome it triggers all the nearby gnomes to join in.

File: code/datums/elements/ai_retaliate.dm (hunk 1 above)
Comment (CapybaraExtravagante) [reply code/datums/elements/ai_retaliate.dm:8]: but maybe @tralezab disagrees with me? :D

Comment (Comxy) [reply code/datums/elements/ai_retaliate.dm:10]: I did it like this since it is essentially the input. I could also try making a child element for it but I'll wait to tralezab to respond.

Comment (tralezab) [reply code/datums/elements/ai_retaliate.dm:10]: Fikou told me of a callback idea I really liked, but this would also make for a good datum ai_behavior as well, keeping the retaliate clean of "optional behavior" which can stack up over time. If @CapybaraExtravagante is happy to help you turn this into an ai behavior, we should go for that. Otherwise, an optional callback to the gnome where it can rouse its friends works. But try ai behavior first!

File: code/datums/elements/ai_retaliate.dm [hunk 3]
```diff
@@ -3,12 +3,18 @@
...
+	var/call_range
 
-/datum/element/ai_retaliate/Attach(datum/target)
+/datum/element/ai_retaliate/Attach(datum/target, input_mob_type = null, input_call_range = 7)
 	. = ..()
 	if(!ismob(target))
 		return ELEMENT_INCOMPATIBLE
 
+	mob_type = input_mob_type
+	call_range = input_call_range
```
Comment (tralezab) [inline code/datums/elements/ai_retaliate.dm:19]: https://github.com/tgstation/tgstation/blob/master/.github/guides/STYLE.md#when-passing-vars-through-new-or-initializes-arguments-use-srcvar

tl;dr: use `src.mob_type = mob_type` instead of `mob_type = input_mob_type`
I mean, this has to change anyways but it's good to point out

**Review by tralezab — CHANGES_REQUESTED** [review]
this pr is pretty awesome

File: code/datums/elements/ai_retaliate.dm (hunk 1 above)
Comment (Comxy) [reply code/datums/elements/ai_retaliate.dm:10]: I used a callback, reason being the AI behavior did weird for me and it having a cool-down to check for each others gnome lists every two seconds or so didn't seem efficient.

Comment (CapybaraExtravagante) [reply code/datums/elements/ai_retaliate.dm]: Thats because you werent supposed to check the other gnomes list but instead only your own list :p

Comment (Comxy) [reply code/datums/elements/ai_retaliate.dm]: I only used that for the callback I made now, back when I used a behavior for it, the gnomes constantly stood still for a second to check for the targeting behavior I think

**Review by tralezab — APPROVED** [review]
Making the retaliate a new component instead of just an optional arg seems very odd to me, but the use of callback is pretty nice and what I cared about personally a lot more.

community-ken-jeong
